### PR TITLE
model: support Step3V

### PR DESCRIPTION
**Pull request by CatherineSue**
## Motivation

This PR adds the support for Step3VModel.

Co-authored-by: Qiaolin-Yu <liin1211@outlook.com>
Co-authored-by: ispobock <ispobaoke@gmail.com>
Co-authored-by: nnnobody-code <nnnobody@foxmail.com>
Co-authored-by: jimpang <jimpang@tencent.com>

## Modifications

- Add Stepv3 model
- Add tool_call_parser `Step3VDetector`
- Add reasoning parser registry for `stepv3`

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/function_call/step3v_detector.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+def parse_arguments(value: str) -> tuple[Any, bool]:
+    """Parse a string value to appropriate type. Returns (parsed_value, success)."""
+    try:
+        try:
+            parsed_value = json.loads(value)
+        except:
+            parsed_value = ast.literal_eval(value)
+        return parsed_value, True
+    except:
+        return value, False
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply python/sglang/srt/function_call/step3v_detector.py]: We can be a bit brute force here and just catch all exceptions.

File: python/sglang/srt/configs/model_config.py [hunk 2]
```diff
@@ -335,10 +335,13 @@ def get_total_num_kv_heads(self) -> int:
             "num_key_value_heads",
             # For ChatGLM:
             "multi_query_group_num",
+            # For Step3
+            "num_attention_groups",
         ]
         for attr in attributes:
             num_kv_heads = getattr(self.hf_text_config, attr, None)
             if num_kv_heads is not None:
+                print("debug attr", attr, num_kv_heads)
```
Comment (CatherineSue) [inline python/sglang/srt/configs/model_config.py]: TODO: remove this

File: python/sglang/srt/configs/__init__.py [hunk 3]
```diff
@@ -14,4 +19,7 @@
     "MultiModalityConfig",
     "KimiVLConfig",
     "MoonViTConfig",
+    "Step3VLConfig",
```
Comment (mickqian) [inline python/sglang/srt/configs/__init__.py:22]: nit: put them in a same file would probably be better

File: python/sglang/srt/models/step3v.py [hunk 4]
```diff
@@ -0,0 +1,1025 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+"""
+Text Model
+"""
+
+
+class Step3TextMLP(nn.Module):
```
Comment (mickqian) [inline python/sglang/srt/models/step3v.py]: duplicate with many other mlps

File: python/sglang/srt/models/step3v.py [hunk 5]
```diff
@@ -0,0 +1,1025 @@
...
+        )
+        self.iter = 0
+
+    def forward(
+        self,
+        positions: torch.Tensor,
+        hidden_states: torch.Tensor,
+        forward_batch: ForwardBatch,
+    ) -> torch.Tensor:
+        self.iter += 1
```
Comment (mickqian) [inline python/sglang/srt/models/step3v.py]: remove this line

File: python/sglang/srt/models/step3v.py [hunk 6]
```diff
@@ -0,0 +1,1025 @@
...
+            padding=1,
+        )
+        self.vit_large_projector = nn.Linear(
+            config.vision_config.output_hidden_size * 2,
+            config.hidden_size,
+            bias=config.projector_bias,
+        )
+
+        self.n_shared_experts = 1
+        # self.num_fused_shared_experts = (
```
Comment (mickqian) [inline python/sglang/srt/models/step3v.py]: please remove this

File: python/sglang/srt/multimodal/processors/step3v.py [hunk 7]
```diff
@@ -0,0 +1,541 @@
...
+        mean = [0.48145466, 0.4578275, 0.40821073]
+        std = [0.26862954, 0.26130258, 0.27577711]
+
+    def preprocess(self, image):
+        return {"pixel_values": self.transform(image).unsqueeze(0)}
+
+    def __call__(self, image):
+        return self.preprocess(image)
+
+    # def process_and_combine_mm_data(
```
Comment (mickqian) [inline python/sglang/srt/multimodal/processors/step3v.py]: nit: remove this

File: test/srt/openai_server/function_call/test_openai_function_calling.py [hunk 8]
```diff
@@ -19,42 +19,43 @@ class TestOpenAIServerFunctionCalling(CustomTestCase):
...
-            cls.base_url,
-            timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
-            api_key=cls.api_key,
-            other_args=[
-                # If your server needs extra parameters to test function calling, please add them here.
-                "--tool-call-parser",
-                "llama3",
-            ],
-        )
+        # cls.process = popen_launch_server(
```
Comment (mickqian) [inline test/srt/openai_server/function_call/test_openai_function_calling.py]: nit

File: test/srt/openai_server/function_call/test_openai_function_calling.py [hunk 9]
```diff
@@ -394,14 +405,16 @@ def test_function_call_strict(self):
...
-            max_tokens=2048,
+            max_tokens=8000,
             messages=messages,
             temperature=0.8,
             top_p=0.8,
             stream=False,
             tools=tools,
         )
 
+        print(response)
```
Comment (mickqian) [inline test/srt/openai_server/function_call/test_openai_function_calling.py]: same as above

File: test/srt/openai_server/function_call/test_openai_function_calling.py [hunk 10]
```diff
@@ -472,6 +485,8 @@ def test_function_call_required(self):
             tool_choice="required",
         )
 
+        print(response)
```
Comment (mickqian) [inline test/srt/openai_server/function_call/test_openai_function_calling.py]: same as above

File: test/srt/openai_server/function_call/test_openai_function_calling.py [hunk 11]
```diff
@@ -494,10 +509,10 @@ def test_function_call_required(self):
             str,
             f"Parameter city should be a string, got: {type(city_value)}",
         )
-        self.assertTrue(
-            "Paris" in city_value or "France" in city_value,
-            f"Parameter city should contain either 'Paris' or 'France', got: {city_value}",
-        )
+        # self.assertTrue(
```
Comment (mickqian) [inline test/srt/openai_server/function_call/test_openai_function_calling.py]: same as above

**Review by mickqian — CHANGES_REQUESTED** [review]

File: python/sglang/srt/managers/template_manager.py [hunk 12]
```diff
@@ -71,31 +71,61 @@ def jinja_template_content_format(self) -> Optional[str]:
...
 
         Args:
             tokenizer_manager: The tokenizer manager instance
-            chat_template_arg: Template name or file path
+            chat_template_arg: Template name, file path, or None to auto-detect
             model_path: Path to the model
         """
-        logger.info(f"Loading chat template: {chat_template_arg}")
+        if chat_template_arg:
+            self._load_explicit_chat_template(tokenizer_manager, chat_template_arg)
```
Comment (CatherineSue) [inline python/sglang/srt/managers/template_manager.py:85]: Test:
Screenshot 2025-07-30 at 11 46 06 PM

File: python/sglang/srt/managers/template_manager.py [hunk 13]
```diff
@@ -71,31 +71,61 @@ def jinja_template_content_format(self) -> Optional[str]:
...
+        else:
+            # Try HuggingFace template first
+            hf_template = self._resolve_hf_chat_template(tokenizer_manager)
+            if hf_template:
+                self._jinja_template_content_format = (
+                    detect_jinja_template_content_format(hf_template)
+                )
+                logger.info(
+                    f"Using default HuggingFace chat template with detected content format: {self._jinja_template_content_format}"
```
Comment (CatherineSue) [inline python/sglang/srt/managers/template_manager.py:94]: Test: 
WechatIMG765
